src/storage/interpreter.rs

```diff
@@ -137,6 +137,34 @@ pub trait InterpreterStorage:
         self.storage::<ContractsState>().remove(&(contract, key))
     }
 
+    /// Fetch a range of values from a key-value mapping in a contract storage.
+    /// Returns the full range requested using optional values in case
+    /// a requested slot is unset.  
+    fn merkle_contract_state_range(
+        &self,
+        id: &ContractId,
+        start_key: &Bytes32,
+        range: Word,
```

Comment:
I can't get what is `range` here and why it is `Word`=)

The storage slots are not required to follow each other. They can be randomly initialized, and in this case, how should it work?

Is it like `count` and after `start_key` we want to fetch the next `count` available slots? If yes, then how can they be empty?

I see how we can return the merkle tree by some merkle root(In this case, we can iterate throw the whole tree, and the code will work with sub-trees). But seems we want something another=)

Reply:
It's similar but not quite exactly the same. It's basically a "fetch all keys between [start_key..start_key + range]" assuming you convert the keys to a u256 or something similar. If there are unset slots within that range, we need to zero out the corresponding memory array in the VM.

Reply:
Hmm, then the usage of `u256` in the signature definitely would make it cleaner=D Or we can pass something like `end: Byte32` or `end: Option<Byte32>'.

What is our limitation for `range`?

From the linked issues, I get that we have two states of the storage slots: set/unset. If the value is zeroed or not zeroed - it is set. If the value is `None` - unset. What is the return data for this method in the next case?

Storage slots:
[Set(1)], Set(2), ..., Set(x - 1), Unset(x), Unset(x + 1), ..., Unset(n)]

What is output `merkle_contract_state_range(start = 1, range = n)`? Is the length of the output vector `n`? Or is it `x - 1`?

Reply:
From the [specs](https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/instruction_set.md#srwq-state-read-sequential-32-byte-slots), it's prescribed as `MEM[$rA, 32 * rD] = STATE[MEM[$rC, 32 * rD]];`.

So this would mean the length of the returned vector should be "range" or `n`, including the start_key.

Since we need to zero out 32bytes slots in memory if that slot is empty, it would be as you say:
> [Set(1)], Set(2), ..., Set(x - 1), Unset(x), Unset(x + 1), ..., Unset(n-1)]

Reply:
> Hmm, then the usage of u256 in the signature definitely would make it cleaner=D Or we can pass something like end: Byte32 or `end: Option'.

We currently don't have a u256 primitive usable in our fuel-types, and without those it would be harder to interact with an "end" style api from the VM perspective.

Reply:
Hmm, okay. Maybe instead of returning `None` for each `Unset` element, we can not return them at all?

If the user requests `start...start+range` and get only some subset, it means he got only `Set` slots, all others are `Unset`. In this case, we can support big `range` values and user can ask for all available slots in the whole storage by `start = 0, range = u256::Max`.

Reply:
The thing is we need to know the exact position of each slot that was unset, and clear the corresponding bytes in `MEM[$rA + 32 * i, 32]`

Reply:
Since we will charge them the same fee regardless of whether the data was set or not, large collections of state like that likely wouldn't fix into a normal tx gas limit either.

Reply:
> The thing is we need to know the exact position of each slot that was unset, and clear the corresponding bytes in MEM[$rA + 32 * i, 32] from the VM memory

I see, because we don't return the length of the return data`MEM[$rA, 32 * rD] = STATE[MEM[$rC, 32 * rD]]` and we expect the array of the same size, we need to zero `Unset` slots.

Suppose the command would support dynamic length return(return length <= `rD`) - the array's length with only `Set` slots. In that case, we could return only `Set` values, and we don't need to charge the user for the processing of `Unset` because we didn't spend any computation resources(In RocksDB all data are sorted by key, `iter_all` will process only available slots). 

But it is another kind of command that we can implement later.

Reply:
yeah rocksdb iterators were my first impression when I originally saw the spec proposal, but that is a different behavior.